diff -up ./ares_parse_srv_reply.c.sigbus ./ares_parse_srv_reply.c
--- ./ares_parse_srv_reply.c.sigbus	2012-03-03 21:30:39.938819989 +0100
+++ ./ares_parse_srv_reply.c	2012-03-03 21:30:54.187689707 +0100
@@ -140,11 +140,11 @@ ares_parse_srv_reply (const unsigned cha
           srv_last = srv_curr;
 
           vptr = aptr;
-          srv_curr->priority = ntohs (*((unsigned short *)vptr));
+          srv_curr->priority = DNS__16BIT(vptr);
           vptr += sizeof(unsigned short);
-          srv_curr->weight = ntohs (*((unsigned short *)vptr));
+          srv_curr->weight = DNS__16BIT(vptr);
           vptr += sizeof(unsigned short);
-          srv_curr->port = ntohs (*((unsigned short *)vptr));
+          srv_curr->port = DNS__16BIT(vptr);
           vptr += sizeof(unsigned short);
 
           status = ares_expand_name (vptr, abuf, alen, &srv_curr->host, &len);
