diff -up ./ares_data.c.break ./ares_data.c
--- ./ares_data.c.break	2012-03-03 19:35:53.201448323 +0100
+++ ./ares_data.c	2012-03-03 19:36:16.177250861 +0100
@@ -123,6 +123,7 @@ void *ares_malloc_data(ares_datatype typ
         ptr->data.addr_node.family = 0;
         memset(&ptr->data.addr_node.addrV6, 0,
           sizeof(ptr->data.addr_node.addrV6));
+        break;
 
       default:
         free(ptr);
